components/p3a_utils/feature_usage.cc

```diff
@@ -0,0 +1,137 @@
+/* Copyright 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/p3a_utils/feature_usage.h"
+
+#include "base/metrics/histogram_functions.h"
+#include "brave/components/p3a_utils/bucket.h"
+#include "brave/components/time_period_storage/monthly_storage.h"
+#include "components/prefs/pref_registry_simple.h"
+#include "components/prefs/pref_service.h"
+
+namespace p3a_utils {
+
+void RegisterFeatureUsagePrefs(PrefRegistrySimple* registry,
+                               const char* first_use_time_pref_name,
+                               const char* last_use_time_pref_name,
+                               const char* used_second_day_pref_name,
+                               const char* days_in_month_used_pref_name) {
+  DCHECK(first_use_time_pref_name);
+  DCHECK(last_use_time_pref_name);
+
+  registry->RegisterTimePref(first_use_time_pref_name, base::Time());
+  registry->RegisterTimePref(last_use_time_pref_name, base::Time());
+
+  if (used_second_day_pref_name != nullptr) {
+    registry->RegisterBooleanPref(used_second_day_pref_name, false);
+  }
+  if (days_in_month_used_pref_name != nullptr) {
+    registry->RegisterListPref(days_in_month_used_pref_name);
+  }
+}
+
+void RecordFeatureUsage(PrefService* prefs,
+                        const char* first_use_time_pref_name,
+                        const char* last_use_time_pref_name) {
+  base::Time now_midnight = base::Time::Now().LocalMidnight();
+  prefs->SetTime(last_use_time_pref_name, now_midnight);
+  if (prefs->GetTime(first_use_time_pref_name).is_null()) {
+    prefs->SetTime(first_use_time_pref_name, now_midnight);
+  }
+}
+
+void MaybeRecordFeatureExistingUsageTimestamp(
+    PrefService* prefs,
+    const char* first_use_time_pref_name,
+    const char* last_use_time_pref_name,
+    base::Time external_last_use_timestamp) {
+  if (!prefs->GetTime(first_use_time_pref_name).is_null() ||
+      external_last_use_timestamp.is_null()) {
+    return;
+  }
+  // If first use time is null and external ts is not,
+  // set the last use time to the external ts so the user does not appear new
+  // in the "new user returning" metric
+  prefs->SetTime(first_use_time_pref_name,
+                 external_last_use_timestamp - base::Days(90));
```

Comment:
why 90?

Reply:
This is to trigger answer = 1 for the NewUserReturning metric, which requires at least 7 days between the first/last use times. I chose 90 days in case we introduce a future feature usage metric which considers a first/last use time delta in the order of months.